ruby/lib/ci/queue/configuration.rb

```diff
@@ -1,81 +1,78 @@
         end
 
-        def load_flaky_tests(path)
-          return [] unless path
-          ::File.readlines(path).map(&:chomp).to_set
-        rescue SystemCallError
-          []
+        def max_duration=(duration)
+          return unless duration
+          circuit_breakers << CircuitBreaker::Timeout.new(duration: duration)
         end
-      end
 
-      def initialize(
-        timeout: 30, build_id: nil, worker_id: nil, max_requeues: 0, requeue_tolerance: 0,
-        namespace: nil, seed: nil, flaky_tests: [], statsd_endpoint: nil, max_consecutive_failures: nil,
-        grind_count: nil, max_duration: nil, failure_file: nil
-      )
-        @circuit_breakers = [CircuitBreaker::Disabled]
-        @build_id = build_id
-        @failure_file = failure_file
-        @flaky_tests = flaky_tests
-        @grind_count = grind_count
-        @max_requeues = max_requeues
-        @namespace = namespace
-        @requeue_tolerance = requeue_tolerance
-        @seed = seed
-        @statsd_endpoint = statsd_endpoint
-        @timeout = timeout
-        @worker_id = worker_id
-        self.max_duration = max_duration
-        self.max_consecutive_failures = max_consecutive_failures
-      end
-
-      def max_consecutive_failures=(max)
-        if max
-          @circuit_breakers << CircuitBreaker.new(max_consecutive_failures: max)
+        def global_max_requeues(tests_count)
+          (tests_count * Float(requeue_tolerance)).ceil
         end
-      end
 
-      def max_duration=(duration)
-        if duration
-          @circuit_breakers << CircuitBreaker::Timeout.new(duration: duration)
+        def flaky?(test_id)
+          flaky_tests.include?(test_id)
         end
       end
 
-      def flaky?(test)
-        @flaky_tests.include?(test.id)
-      end
+      attr_reader :config
+
+      def initialize(env, argv)
+        env_config = CI::Queue::ParseEnv.parse(env)
+        argv_config = CI::Queue::ParseArgv.new(argv).config
 
-      def seed
-        @seed || build_id
+        @config = Config.new()
+        @config.worker_id = argv_config.worker_id || env_config.worker_id
+        @config.grind_list = argv_config.grind_list
+        @config.grind_count = argv_config.grind_count
+        @config.statsd_endpoint = env_config.statsd_endpoint
+        @config.failing_test = argv_config.failing_test
+        @config.failure_file = argv_config.failure_file
+        @config.circuit_breakers = [CircuitBreaker::Disabled]
+        @config.max_requeues = argv_config.max_requeues || 0
+        @config.requeue_tolerance = argv_config.requeue_tolerance || 0
+        @config.queue_url = argv_config.queue_url || env_config.queue_url
+        @config.flaky_tests = env_config.flaky_tests || []
+        @config.timeout = argv_config.timeout || 30
+        build_id = fetch_build_id(argv_config.namespace, env_config.build_id, argv_config.build_id)
+        @config.build_id = build_id
+        @config.seed = argv_config.seed || env_config.seed || build_id
```

Comment:
We are duplicating logic here from Argv and Env config. It might be good to just access them directly from those objects.